aws/resource_aws_servicecatalog_provisioning_artifact.go

```diff
@@ -0,0 +1,297 @@
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
+)
+
+func resourceAwsServiceCatalogProvisioningArtifact() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsServiceCatalogProvisioningArtifactCreate,
+		Read:   resourceAwsServiceCatalogProvisioningArtifactRead,
+		Update: resourceAwsServiceCatalogProvisioningArtifactUpdate,
+		Delete: resourceAwsServiceCatalogProvisioningArtifactDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"accept_language": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "en",
+				ValidateFunc: validation.StringInSlice(tfservicecatalog.AcceptLanguage_Values(), false),
+			},
+			"active": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"created_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"disable_template_validation": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+				ForceNew: true,
+			},
+			"guidance": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      servicecatalog.ProvisioningArtifactGuidanceDefault,
+				ValidateFunc: validation.StringInSlice(servicecatalog.ProvisioningArtifactGuidance_Values(), false),
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"product_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"template_physical_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ExactlyOneOf: []string{
+					"template_url",
+					"template_physical_id",
+				},
+			},
+			"template_url": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ExactlyOneOf: []string{
+					"template_url",
+					"template_physical_id",
+				},
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(servicecatalog.ProvisioningArtifactType_Values(), false),
+			},
+		},
+	}
+}
+
+func resourceAwsServiceCatalogProvisioningArtifactCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+
+	parameters := make(map[string]interface{})
+	parameters["description"] = d.Get("description")
+	parameters["disable_template_validation"] = d.Get("disable_template_validation")
+	parameters["name"] = d.Get("name")
+	parameters["template_physical_id"] = d.Get("template_physical_id")
+	parameters["template_url"] = d.Get("template_url")
+	parameters["type"] = d.Get("type")
+
+	input := &servicecatalog.CreateProvisioningArtifactInput{
+		IdempotencyToken: aws.String(resource.UniqueId()),
+		Parameters:       expandProvisioningArtifactParameters(parameters),
```

Comment:
just curious about this usage: why not pass *schema.ResourceData directly to `expandProvisioningArtifactParameters()`?

Reply:
not sure i follow. the expander takes a `map[string]interface{}`. is there a better way to get root data as a `map[string]interface{}`? that would be cool!